src/main/scala/tilelink/AddressAdjuster.scala

```diff
@@ -7,203 +7,315 @@ import chisel3.util._
-        require (!l.supportsArithmetic || r.supportsArithmetic, s"Device ${l.name} (${l.address}) loses Arithmetic support because ${r.name} does not support it")
-        require (!l.supportsLogical    || r.supportsLogical,    s"Device ${l.name} (${l.address}) loses Logical support because ${r.name} does not support it")
-        require (!l.supportsGet        || r.supportsGet,        s"Device ${l.name} (${l.address}) loses Get support because ${r.name} does not support it")
-        require (!l.supportsPutFull    || r.supportsPutFull,    s"Device ${l.name} (${l.address}) loses PutFull support because ${r.name} does not support it")
-        require (!l.supportsPutPartial || r.supportsPutPartial, s"Device ${l.name} (${l.address}) loses PutPartial support because ${r.name} does not support it")
-        require (!l.supportsHint       || r.supportsHint,       s"Device ${l.name} (${l.address}) loses Hint support because ${r.name} does not support it")
-        l.copy(
-          // take the 0 setting as default for DTS output
-          address            = AddressSet.unify(la) ++ (if (l == errorDev) holes else Nil),
-          regionType         = r.regionType,
-          executable         = r.executable,
-          supportsAcquireT   = r.supportsAcquireT,
-          supportsAcquireB   = r.supportsAcquireB,
-          supportsArithmetic = r.supportsArithmetic,
-          supportsLogical    = r.supportsLogical,
-          supportsGet        = r.supportsGet,
-          supportsPutFull    = r.supportsPutFull,
-          supportsPutPartial = r.supportsPutPartial,
-          supportsHint       = r.supportsHint,
-          mayDenyGet         = r.mayDenyGet,
-          mayDenyPut         = r.mayDenyPut,
-          alwaysGrantsT      = r.alwaysGrantsT,
-          fifoId             = fifoId)
-      }
+        // For address adjustment to be possible, we have to calculate some specific things about the downstream addess map:
 
-      val newRemote = ids.tail.zipWithIndex.flatMap { case (id, i) => remote.managers.map { r =>
-        r.copy(
-          address = AddressSet.unify(r.address.flatMap(_.intersect(AddressSet(id, ~mask)))),
-          fifoId = Some(i+1))
-      } }
+        // Find the downstream error device
+        val errorDevs = local.managers.filter(_.nodePath.last.lazyModule.className == "TLError")
+        require (!errorDevs.isEmpty, s"There is no TLError reachable from ${name}. One must be instantiated.")
+        val errorDev = errorDevs.head
 
-      local.copy(
-        managers   = newLocal ++ newRemote,
-        endSinkId  = local.endSinkId + remote.endSinkId,
-        minLatency = local.minLatency min remote.minLatency)
+        // Find all the holes in local routing
+        val holes = {
+          val ra = masked(adjustableRemoteManagers.flatMap(_.address))
+          val la = masked(adjustableLocalManagers .flatMap(_.address))
+          la.foldLeft(ra) { case (holes, la) => holes.flatMap(_.subtract(la)) }
+        }
+
```

Comment:
I'm concerned that we don't check that the adjustableRemoteManagers.flatMap completely cover the adjustableRegion. On the other hand, we would like adjustableRegion to be optional.

Reply:
Maybe run address unification on adjustableRemoteManagers.flatMap(_.address) and then require it to be contiguous?

Reply:
Never mind. This already suffices:
          val container = adjustableRemoteManagers.find { r => l.address.forall { la => r.address.exists(_.contains(la)) } }